rad/prelude/patterns.rad

```diff
@@ -0,0 +1,192 @@
+                :conflicts (insert k [(lookup k merge) v] (lookup :conflicts acc))}
+               (insert :merge (insert k v merge) acc)))
+           {:merge m
+            :conflicts {}}
+           (seq n))))
+
+(def non-linear-merge
+  "The bindings merge strategy for non-linear patterns. Use this function to merge
+  bindings returned by sub-patterns if you want your pattern to be non-linear."
+  (fn [m n]
+    (def excl-merge (exclusive-dict-merge m n))
+    (def cs (lookup :conflicts excl-merge))
+    (def merge (lookup :merge excl-merge))
+    (if (eq? cs {})
+      [:Just merge]
+      (maybe-foldlM (fn [acc kv]
+                      (def v (nth 1 kv))
+                      (def v0 (nth 0 v))
+                      (if (eq? v0 (nth 1 v))
+                        [:Just (insert (nth 0 kv) v0 acc)]
+                        :Nothing))
+                    merge
+                    (seq cs)))))
+(def _
+  "The wildcard pattern."
+  (fn [v] [:Just {}]))
+
+(def-rec match-pat
+  "The pattern matching dispatch function. This function defines how patterns
+  are treated in `match` expressions. Atoms are treated as bindings. Numbers,
+  keywords and strings are constant patterns. Dicts of patterns match dicts
+  whose values at those keys match those patterns. Vectors of patterns match
+  vectors of the same length, pairing the patterns and elements by index."
+  (fn [pat v]
+    (def /vec
+      (fn [pats]
+        (fn [v]
+          (if (eq? (length v) (length pat))
+            (maybe-foldlM
+             (fn [acc pair]
+               (maybe->>= (match-pat (nth 0 pair) (nth 1 pair))
+                          (fn [bs] (non-linear-merge acc bs))))
+             {}
+             (zip pats v))
+            :Nothing))))
+    (def /dict
+      (fn [keys-pat]
+        (fn [d]
+          (maybe-foldlM
+           (fn [acc kv]
+             (def k (nth 0 kv))
+             (def pat (nth 1 kv))
+             (if (member? k d)
+               (maybe->>= (match-pat pat (lookup k d))
+                          (fn [bs] (non-linear-merge acc bs)))
+               :Nothing))
+           {}
+           (seq keys-pat)))))
+    (cond
+      (atom? pat)
+      [:Just {pat v}]
+      (some [(number? pat) (keyword? pat) (string? pat)])
+      (if (eq? pat v) [:Just {}] :Nothing)
+      (dict? pat)
+      ((/dict pat) v)
+      (vector? pat)
+      ((/vec pat) v)
+      :else
+      (pat v))))
+
+(:test "basic pattern matching"
+       [ (match 42 12 :not-42 'x x) ==> 42 ]
+       [ (match 42 12 :not-42 42 :yes) ==> :yes ]
+       [ (match :fourty-two :not-42 :no :fourty-two :yes) ==> :yes ]
+       [ (match "42" "?" :no "42" :yes) ==> :yes ]
+       [ (match {:a :a :b :b} {:a 'a :b 'b} [a b]) ==> [:a :b] ]
+       [ (match [1 2 3] ['x _ 'y] [x y]) ==> [1 3] ]
+       [ (match {:a 1 :b [1 2 3]}
+                42 :no
+                ({:a _ :b [_ 'n _]} | (eq? n 2)) :yes
+                _ :no) ==> :yes ]
+       [ (match [1 2]
```

Comment:
What about when the thing being matched is an identifier? Shouldn't that also match a (double-quoted here) identifier?

Reply:
That would involve adding a case for `(quote x)` in the dispatch function, which I'm not sure is a good idea, because we might want to use lists for matching lists or something. People can write a pattern for matching atoms if they want.